internal/pkg/vm/internal/runtime/runtime.go

```diff
@@ -8,27 +8,142 @@ import (
+type initInvocationContext interface {
+	InvocationContext
+	// Review: for @spec, added an error that is not yet on the spec but should be
+	// Note: @jzimmerman: "Agreed, this should support an error return."
+	CreateActor(cid cid.Cid, addr address.Address, constructorParams MethodParams) error
+}
 
+// minerInvocationContext has some special sauce for the miner.
+// Review: for @go-filecoin, do we want custom interfaces for the miners instead of grouping a bunch fo methods on the InvocationContext?
+type minerInvocationContext interface {
 	Verifier() verification.Verifier
 }
 
+// ActorStateHandler handles the actor state, allowing actors to lock on the state.
+type ActorStateHandler interface {
+	// Take returns the `Cid` for the state and locks the state.
+	//
+	// Any futre calls to `Take` by other instances will block until `Release` or `UpdateRelease` is called.
+	// Review: for @spec, the spec impl of `Take` is not doing any blocking, just panic'ing on the second call on the SAME object instance (this objects are not kept around, so it is always a new instance..).
+	Take() cid.Cid
+	// UpdateRelease updates the actor state and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	UpdateRelease(newStateCID ActorSubstateCID)
+	// Release asserts that the state has not changed and releases the lock on it.
+	//
+	// No future calls to `Take` are allowed on this object.
+	// Review: for @spec, this method can currently be called without having called `Take` first.
+	// Review: for @spec, why is this method required?
+	Release(checkStateCID ActorSubstateCID)
+}
+
+// IpldStorage2 defines the storage module exposed to actors.
+type IpldStorage2 interface {
+	Put(interface{}) (cid.Cid, error)
+	Get(cid.Cid, interface{}) error
+}
+
+// CallerPattern checks if the caller matches the pattern.
+type CallerPattern interface {
+	IsMatch() bool
+}
+
+// Epoch is the current chain epoch.
+// Dragons: this is just to match names on the VM, Height got renamed to Epoch on the spec
+// TODO: rename types.BlockHeight to block.Epoch (issue: ???)
+type Epoch types.BlockHeight
+
+// Randomness is a string of random bytes
+// Note: placeholder for matching newtypes with the spec.
+type Randomness []byte
+
+// InvocInput are the params to invoke a method in an Actor.
+type InvocInput struct {
+    To      address.Address
+    Method  types.MethodID
+    Params  MethodParams
+    Value   types.AttoFIL
+}
+
+// MethodParam is the parameter to an Actor method already encoded for execution in the VM.
+type MethodParam []byte
+// MethodParams is a list of `MethodParam` to be passed into an Actor method.
+type MethodParams []MethodParam
+
+// Abort will stop the VM execution and return an abort error to the caller.
+// Review: @go-filecoin, do we want to handle this with panics?
```

Comment:
I am pretty torn on this. Our code will line up better with the spec if we do use a panic mechanism. If we do, we should wrap the argument to `panic` in a struct so that a `recover` at the top of the call tree can distinguish an explicit call abort (which all nodes should reproduce) from an unreproducible environmental failure (e.g. out of memory), the latter of which should crash the node.

It's my understanding that an aborted call will not result in a receipt. It does consume gas and increment the caller's sequence number, though. I.e. equivalent to running out of gas. I get confused every time I think about this tho, so could be wrong.

We need also to distinguish between this and an "application" failure that "successfully" returns a non-zero exit code.

Reply:
I don't particularly like using panics for control flow.

> It's my understanding that an aborted call will not result in a receipt. It does consume gas and increment the caller's sequence number, though. I.e. equivalent to running out of gas. I get confused every time I think about this tho, so could be wrong.

The abort consuming gas has some implications on how we manage gas and handle the panic. 
The gas consumed by a function 3 levels deep which panics, needs to be accounted when the top level call traps the panic.

> We need also to distinguish between this and an "application" failure that "successfully" returns a non-zero exit code.

The abort already has a unique error status code (code=8).

Reply:
In case it's helpful, there's already a reference implementation of this error struct mechanism (`RuntimeError`) in the spec: `systems/filecoin_vm/runtime/runtime.go`, see in particular the `_catchRuntimeErrors` function.

(Also note I have a pending PR to remove the `MethodPanic` case in that function, per our earlier discussion, since we shouldn't be intercepting nondeterministic/nonreproducible panics here.)

Reply:
(Side note: @icorderi, for what it's worth, I also don't like using panics for control flow in general, but in Go we don't have a robust error-handling mechanism such as option types or Rust's `try!` macros, so I didn't see a good alternative.)

Reply:
I may have been wrong about aborted calls not resulting in a receipt. I was talking about old behaviour.

Re gas: I have heard somewhere that Eth consumes the entire gas limit for a message when it aborts, rather than attempt consensus about the cost (and also disincentivize malformed messages). So the accounting may be simple. However we need this to be clearly specified and to check it doesn't mess up incentives with block validation rules etc (e.g. miners including messages out of sequence order in order to collect max gas).